criu/cr-check.c

```diff
@@ -1666,36 +1676,43 @@ static int pr_set_dumpable(int value)
 
 int check_caps(void)
 {
-	struct proc_status_creds creds;
-	int exit_code = -1;
-
-	if (parse_pid_status(PROC_SELF, &creds.s, NULL))
+	/* Read out effective capabilities and store in opts.cap_eff. */
+	if (set_opts_cap_eff())
 		goto out;
 
-	memcpy(&opts.cap_eff, &creds.cap_eff, sizeof(u32) * PROC_CAP_SIZE);
-
+	/*
+	 * No matter if running as root or not. CRIU always needs
+	 * at least these capabilities.
+	 */
 	if (!has_cap_checkpoint_restore(opts.cap_eff))
 		goto out;
 
 	/* For some things we need to know if we are running as root. */
 	opts.uid = geteuid();
 
-	if (opts.uid) {
-		/*
-		 * At his point we know we are running as non-root with the necessary
-		 * capabilities available. Now we have to make the process dumpable
-		 * so that /proc/self is not owned by root.
-		 */
-		if (pr_set_dumpable(1))
-			return -1;
+	if (!opts.uid) {
+		/* CRIU is running as root. No further checks are necessary. */
+		return 0;
 	}
 
-	exit_code = 0;
-out:
-	if (exit_code) {
-		pr_msg("CRIU needs to have the CAP_SYS_ADMIN or the CAP_CHECKPOINT_RESTORE capability: \n");
-		pr_msg("setcap cap_checkpoint_restore+eip %s\n", opts.argv_0);
+	if (!opts.unprivileged) {
+		pr_msg("Running as non-root requires '--unprivileged'\n");
+		pr_msg("Please consult the documentation for limitations when running as non-root\n");
+		return -1;
 	}
 
-	return exit_code;
+	/*
+	 * At his point we know we are running as non-root with the necessary
+	 * capabilities available. Now we have to make the process dumpable
+	 * so that /proc/self is not owned by root.
+	 */
+	if (pr_set_dumpable(1))
```

Comment:
hmm, can it have any security implications? Will a user be able to trace a criu process?

Reply:
I looked into this a bit. It's here because CRIU will look at its own `/proc/self/pagemap` when initializing `kdat.pmap` and `kdat.zero_page_pfn`. It can possibly be changed to look at target process' pagemap (or parent's pagemap for the criu-check case), but its a little tricky because we need to find a present page to check if PFNs are available or only flags. The current code uses an easily available stack address to find a present page.

As far as security implications go I don't think there is anything sensitive in the CRIU process itself that will be exposed by tracing or core dumps, and the target process in unprivileged mode will be itself be unprivileged (meaning user can already trace/dump it), so we are not exposing anything new there.